node: topologymgr: Graduate Kubelet Topology Manager to GA

swatisehgal opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature


#### What this PR does / why we need it:
Captures changes to graduate Kubelet Topology Manager to GA.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Issue: [#693](https://github.com/kubernetes/enhancements/issues/693)

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Graduate Kubelet Topology Manager to GA.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
KEP: https://github.com/kubernetes/enhancements/pull/3745
```

swatisehgal committed:
node: topologymgr: Graduate Kubelet Topology Manager to GA

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: topologymgr: GA graduation implies Feature Gate is ON by default

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

bart0sh commented:
/triage accepted
/priority important-soon
/assign @SergeyKanzhelev @klueska @mrunalp

swatisehgal commented:
/cc @ffromani

ffromani reviewed (commented):
minor comments inside.
LGTM once the dependencies are merged.

ffromani commented:
ok, according to https://github.com/kubernetes/enhancements/issues/693 this seems the last PR needed (and this seems correct to me). So let's discuss the inline comments and should be fine.

swatisehgal commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -289,20 +289,15 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
-	} else {
-		cm.topologyManager = topologymanager.NewFakeManager()
+	if err != nil {
+		return nil, err
```
Yes, it is.

ffromani replied:
great, then no further action is needed on this front.

swatisehgal commented on pkg/kubelet/cm/internal_container_lifecycle.go:
```diff
@@ -60,11 +57,9 @@ func (i *internalContainerLifecycleImpl) PreStopContainer(containerID string) er
-		if err != nil {
-			return err
-		}
+	err := i.topologyManager.RemoveContainer(containerID)
```
Good point, will return the error directly.

ffromani commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -289,20 +289,15 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
-		}
+	cm.topologyManager, err = topologymanager.NewManager(
+		machineInfo.Topology,
+		nodeConfig.ExperimentalTopologyManagerPolicy,
+		nodeConfig.ExperimentalTopologyManagerScope,
+		nodeConfig.ExperimentalTopologyManagerPolicyOptions,
```
sorry for late notice: should we rename these fields dropping the `Experimental` prefix? I think we should

/cc @SergeyKanzhelev

swatisehgal replied:
No worries, I can add a commit for this.

ffromani reviewed (commented):
I apologize for the late notice, but it seems we missed a trivial rename. Besides that, everything else LGTM

If you agree with the rename, you can either push a separate commit with the rename or amend the existing last one

swatisehgal committed:
node: topologymgr: trivial internal variable renaming

Since Topology manager is graduating to GA, we remove
internal configuration variable names with `Experimental`
prefix.

There is no expected change in behavior, only trival
variable renaming.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

ffromani commented:
/lgtm

thanks!

ffromani commented:
/retest

 pull-kubernetes-e2e-kind-ipv6 failures can only be flakes considering the content of this PR

ffromani commented:
/retest

```
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/test/integration: TestChangeCRD expand_more
```
this looks like a flake

SergeyKanzhelev reviewed (approved):
/lgtm

swatisehgal commented:
/assign @klueska @mrunalp 
for approval on kubelet changes

/assign @dims @derekwaynecarr
for features approval

swatisehgal committed:
node: topologymgr: Remove ResourceAllocator as TM is always enabled

With Topology Manager enabled by default, we no longer need
`resourceAllocator` as Topology Manager serves as the main
PodAdmitHandler completely responsible for admission check
based on hints received from the hintProviders and the
subsequent allocation of the corresponding resources to a
pod as can be seen here:
https://github.com/kubernetes/kubernetes/blob/v1.26.0/pkg/kubelet/cm/topologymanager/scope.go#L150

With regard to DRA, the passing of `cm.draManager` into
resourceAllocator seems redundant as no admission checks
(and allocation of resources handled by DRA) is taking place
in `Admit` method of resourceAllocator. DRA has a completely
different model to the rest of the resource managers where
pod is only scheduled on a node once resources are reserved
for it. Because of this, admission checks or waiting for
resources to be provisioned after the pod has been scheduled
on the node is not required.

Before making the above change, it was verified that DRA Manager
is instantiated in `NewContainerManager`:
https://github.com/kubernetes/kubernetes/blob/v1.26.0/pkg/kubelet/cm/container_manager_linux.go#L318

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/cc @pohly @bart0sh 
for input on changes related to removal of `resourceAllocator` as `cm.draManager` was passed into it.

klueska commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -683,50 +677,7 @@ func (cm *containerManagerImpl) UpdatePluginResources(node *schedulerframework.N
-	cpuManager    cpumanager.Manager
-	memoryManager memorymanager.Manager
-	deviceManager devicemanager.Manager
-	draManager    dra.Manager
```
It was an oversight not to remove this from here in the first place because allocation was moved out of the pod admission loop and into container start before the DRA PR ever merged. This is leftover from intermediate changes that never merged.

swatisehgal replied:
Thanks @klueska for clarifying this!

klueska reviewed (commented):
Did a quick pass, including addressing your comment about the draManager in the (now removed) resourceAllocator.

/lgtm
/approve

swatisehgal commented:
/label api-review
due to changes made in `apis/config/validation` .

SergeyKanzhelev commented on cmd/kubelet/app/server.go:
```diff
@@ -711,13 +711,11 @@ func run(ctx context.Context, s *options.KubeletServer, kubeDeps *kubelet.Depend
+		if utilfeature.DefaultFeatureGate.Enabled(features.TopologyManagerPolicyOptions) {
+			topologyManagerPolicyOptions = s.TopologyManagerPolicyOptions
+		} else if s.TopologyManagerPolicyOptions != nil {
+			return fmt.Errorf("topology manager policy options %v require feature gates %q, %q enabled",
+				s.TopologyManagerPolicyOptions, features.TopologyManager, features.TopologyManagerPolicyOptions)
```
why it still refers to `features.TopologyManager`?

```suggestion
			return fmt.Errorf("topology manager policy options %v require feature gate  %q enabled",
				s.TopologyManagerPolicyOptions, features.TopologyManagerPolicyOptions)
```

swatisehgal replied:
Good point, will remove.

SergeyKanzhelev commented on pkg/kubelet/cm/container_manager.go:
```diff
@@ -145,18 +145,18 @@ type NodeConfig struct {
+	EnforceCPULimits                        bool
+	CPUCFSQuotaPeriod                       time.Duration
+	TopologyManagerPolicy                   string
+	TopologyManagerPolicyOptions            map[string]string
```
this still requires `TopologyManagerPolicyOptions` gate? Should it stay Experimental?

swatisehgal replied:
Yes it does. Rethinking this, I believe it makes sense to keep the experimental prefix for this as it is still an alpha feature thereby requiring a feature gate to enable it. Once the feature graduates to Beta, we can remove the prefix.

ffromani commented on cmd/kubelet/app/server.go:
```diff
@@ -749,8 +749,8 @@ func run(ctx context.Context, s *options.KubeletServer, kubeDeps *kubelet.Depend
-				ExperimentalTopologyManagerScope:         s.TopologyManagerScope,
+				TopologyManagerPolicy:                    s.TopologyManagerPolicy,
+				TopologyManagerScope:                     s.TopologyManagerScope,
 				ExperimentalTopologyManagerPolicyOptions: topologyManagerPolicyOptions,
```
yes, this seems the tight change at this point in time. ExperimentalTopologyManagerPolicyOptions is governed by a different FG.

ffromani reviewed (commented):
/lgtm

SergeyKanzhelev commented on pkg/features/kube_features.go:
```diff
@@ -757,9 +757,10 @@ const (
 	// Enables topology aware hints for EndpointSlices
 	TopologyAwareHints featuregate.Feature = "TopologyAwareHints"
 
-	// owner: @lmdaly
```
Also need to update types.go - remove wording `requires the TopologyManager feature gate to be enabled`

swatisehgal replied:
Sure, will do now. Thanks!

SergeyKanzhelev commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -199,8 +199,6 @@ func configureTopologyManagerInKubelet(oldCfg *kubeletconfig.KubeletConfiguratio
 		newCfg.FeatureGates = make(map[string]bool)
 	}
 
-	newCfg.FeatureGates["TopologyManager"] = true
```
I think tests do not need the tags `[Feature:TopologyManager][NodeFeature:TopologyManager]` any longer

swatisehgal replied:
A tag such as this can be useful if someone want to run just topology manager node e2e tests e.g. for debugging purposes. I will remove `[Feature:TopologyManager]` and we can keep just `[NodeFeature:TopologyManager]`.

swatisehgal committed:
node: topologymgr: remove comments with feature gate references

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: topologymgr: update node e2e test tags

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: topologymgr: update autogenerated code

Changes committed after running:
`./hack/update-codegen.sh`

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

liggitt commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -289,20 +288,15 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
-		if err != nil {
-			return nil, err
-		}
+	cm.topologyManager, err = topologymanager.NewManager(
```
looking inside the NewManager() impl, should the calls to NewPolicyOptions and NewNUMAInfo calls be skipped if the policy is PolicyNone? Currently, it looks like we can fail to start kubelets if we can't discover NUMA topology, even though that won't be used at all if the policy is PolicyNone

ffromani replied:
This is a valid point. Previously users had the chance to exclude this flow with the FG, this option is now being removed. My 2c: this predates the GA graduation and could require some code massaging (e.g. not a simple if) so IMO is best addressed with a followup PR, which should be tracked (but probably not blocking) in the GA-graduation issue.  I'll let @klueska (cc @SergeyKanzhelev ) make the final call.

SergeyKanzhelev replied:
https://github.com/kubernetes/kubernetes/issues/116391

ffromani replied:
thanks @SergeyKanzhelev !

SergeyKanzhelev reviewed (approved):
/lgtm

Created an issue to follow up and minimize the number of calls needed on initialization when policy is none. Should not block GA

SergeyKanzhelev commented:
/retest

pacoxu commented:
/retest-required
/skip
